crates/bevy_ecs/src/world/error.rs

```diff
@@ -2,11 +2,22 @@
 
 use thiserror::Error;
 
-use crate::schedule::InternedScheduleLabel;
+use crate::{component::ComponentId, schedule::InternedScheduleLabel};
 
 /// The error type returned by [`World::try_run_schedule`] if the provided schedule does not exist.
 ///
 /// [`World::try_run_schedule`]: crate::world::World::try_run_schedule
 #[derive(Error, Debug)]
 #[error("The schedule with the label {0:?} was not found.")]
 pub struct TryRunScheduleError(pub InternedScheduleLabel);
+
+/// An error that occurs when dynamically retrieving components from an entity.
+#[derive(Error, Debug, Clone, Copy, PartialEq, Eq)]
+pub enum EntityComponentError {
+    /// The component with the given [`ComponentId`] does not exist on the entity.
+    #[error("The component with ID {0:?} does not exist on the entity.")]
+    MissingComponent(ComponentId),
+    /// The component with the given [`ComponentId`] was requested mutably more than once.
+    #[error("The component with ID {0:?} was requested mutably more than once.")]
+    AliasedMutability(ComponentId),
```

Comment:
I think mutable + immutable refs is also invalid access

Reply:
True, but with the current API it's not possible to request both mutable and immutable references at the same time. You can only receive either all mutable references, or all immutable references.